### PR TITLE
Update ECR_URL to match localhost repo

### PR DESCRIPTION
**Pull request by sleepwithcoffee**
#### Which issue(s) does this change fix?

#3292 

#### Why is this change necessary?

To make SAM compatible with LocalStack ECR

#### How does it address the issue?

Update ECR_URL regex pattern to be able to match with either AWS hostname or localhost (with or without port) ot `127.0.0.1` (with or without port)

#### What side effects does this change have?

This is an additional add-in, it should not change the current behaviour

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [x] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [x] Write/update integration tests
- [x] Write/update functional tests if needed
- [x] `make pr` passes
- [x] `make update-reproducible-reqs` if dependencies were changed -- no dependencies changed
- [x] Write documentation -- no documentation involved

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

Regarding unit tests & function tests, I skimmed through the source code and realized that function `is_ecr_url()` itself is not covered by any test(?)

So I used the follow script to perform my own testing (inspired by the original comment under #3292 ):

```python
"""
Regular Expressions for Resources.
"""
HOSTNAME_ECR_AWS = r"([a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?"
HOSTNAME_LOCALHOST = r"localhost(:\d{1,5})?"
HOSTNAME_127_0_0_1 = r"127\.0\.0\.1(:\d{1,5})?"
ECR_URL = (
    rf"^({HOSTNAME_ECR_AWS}|{HOSTNAME_LOCALHOST}|{HOSTNAME_127_0_0_1})\/"
    r"(?:[a-z0-9]+(?:[._-][a-z0-9]+)*/)*[a-z0-9]+(?:[._-][a-z0-9]+)*"
)

"""
ECR Packaging Utils
"""
import re

def is_ecr_url(url: str) -> bool:
    return bool(re.match(ECR_URL, url)) if url else False

test_cases = [
    '000000000000.dkr.ecr.eu-west-1.amazonaws.com/my-repo',
    '000000000000.dkr.ecr.eu-west-1.amazonaws.com/a/longer/path/my-repo',
    'localhost/my-repo',
    'localhost:4566/my-repo',
    '127.0.0.1/my-repo',
    '127.0.0.1:1234/my-repo',
    '880954911558.dkr.ecr.us-west-2.amazonaws.com/everflow-app',
    '000000000000.dkr.ecr.eu-west-1.amazonaws.com' # only hostname, no repo
    '000000000000.invalid.amazonaws.com/my-repo',
    'localhost:4566',
    'localhost:123456/my-repo', # port name too long
    'localhost:abc/my-repo', # bad port name
    ':4566',
    'somethingstring',
    ':',
    '/my-repo'
]

for url in test_cases:
    print(is_ecr_url(url), url)

```

Output:
```
True 000000000000.dkr.ecr.eu-west-1.amazonaws.com/my-repo
True 000000000000.dkr.ecr.eu-west-1.amazonaws.com/a/longer/path/my-repo
True localhost/my-repo
True localhost:4566/my-repo
True 127.0.0.1/my-repo
True 127.0.0.1:1234/my-repo
True 880954911558.dkr.ecr.us-west-2.amazonaws.com/everflow-app
False 000000000000.dkr.ecr.eu-west-1.amazonaws.com000000000000.invalid.amazonaws.com/my-repo
False localhost:4566
False localhost:123456/my-repo
False localhost:abc/my-repo
False :4566
False somethingstring
False :
False /my-repo
```

### CODE REVIEW

**Review by moelasmar — COMMENTED** [review]
I found that we do not have any unit test cases that cover the function `is_ecr_url` in module `samcli/lib/package/ecr_utils.py` .. could you please the unit test cases that cover the old and new code ?

File: samcli/lib/package/regexpr.py [hunk 1]
```diff
@@ -2,7 +2,10 @@
 Regular Expressions for Resources.
 """
 # Regex derived from https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_Repository.html
+HOSTNAME_ECR_AWS = r"([a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?"
```
[the remark this answers is not in the document]
Comment (sleepwithcoffee) [reply samcli/lib/package/regexpr.py]: done, however, it's worth to note that non-capturing and capturing groups would yield the same outcome under the function `is_ecr_url()`

**Review by moelasmar — APPROVED** [review]

**Comment by moelasmar** [comment]
> btw, I dont think separting `ECR_URL` and `is_ecr_url()` is a good idea since `ECR_URL` is not used elsewhere anyway

I agree, can you refactor it?

**Comment by sleepwithcoffee** [comment]
updated & ran `make format` to reformat files

**Review by mildaniel — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**